compliance/virtualmachines/relay/provider/provider.go

```diff
@@ -0,0 +1,245 @@
+// reports to the relay for sending to Sensor.
+package provider
+
+import (
+	"context"
+	"net"
+	"strconv"
+	"time"
+
+	"github.com/pkg/errors"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/stackrox/rox/compliance/virtualmachines/relay/connutil"
+	"github.com/stackrox/rox/compliance/virtualmachines/relay/metrics"
+	"github.com/stackrox/rox/compliance/virtualmachines/relay/vsock"
+	v1 "github.com/stackrox/rox/generated/internalapi/virtualmachine/v1"
+	"github.com/stackrox/rox/pkg/env"
+	"github.com/stackrox/rox/pkg/logging"
+	"golang.org/x/sync/semaphore"
+	"google.golang.org/protobuf/proto"
+)
+
+var log = logging.LoggerForModule()
+
+// ReportProvider manages the vsock server and produces validated reports.
+type ReportProvider interface {
+	// Start begins accepting connections and returns a channel of validated reports.
+	// The channel will be closed when the context is cancelled or an unrecoverable error occurs.
+	Start(ctx context.Context) (<-chan *v1.IndexReport, error)
+}
+
+type providerImpl struct {
+	listener              net.Listener
+	semaphore             *semaphore.Weighted
+	semaphoreTimeout      time.Duration
+	connectionReadTimeout time.Duration
+	waitAfterFailedAccept time.Duration
+	maxSizeBytes          int
+}
+
+var _ ReportProvider = (*providerImpl)(nil)
+
+// New creates a ReportProvider with a vsock listener.
+// Concurrency limits are read from env vars VirtualMachinesMaxConcurrentVsockConnections
+// and VirtualMachinesConcurrencyTimeout.
+func New() (ReportProvider, error) {
+	listener, err := vsock.NewListener()
+	if err != nil {
+		return nil, errors.Wrap(err, "creating vsock listener")
+	}
+
+	maxConcurrentConnections := env.VirtualMachinesMaxConcurrentVsockConnections.IntegerSetting()
+	semaphoreTimeout := env.VirtualMachinesConcurrencyTimeout.DurationSetting()
+	maxSizeBytes := env.VirtualMachinesVsockConnMaxSizeKB.IntegerSetting() * 1024
+
+	return &providerImpl{
+		listener:              listener,
+		semaphore:             semaphore.NewWeighted(int64(maxConcurrentConnections)),
+		semaphoreTimeout:      semaphoreTimeout,
+		connectionReadTimeout: 10 * time.Second,
+		waitAfterFailedAccept: time.Second,
+		maxSizeBytes:          maxSizeBytes,
+	}, nil
+}
+
+// Start begins accepting vsock connections and returns a channel of validated reports.
+// The provider spawns goroutines to handle each connection concurrently (up to the
+// configured limit). Reports are validated before being sent to the channel.
+func (p *providerImpl) Start(ctx context.Context) (<-chan *v1.IndexReport, error) {
+	log.Info("Starting report provider")
+
+	if p.listener == nil {
+		return nil, errors.New("listener is nil")
+	}
+
+	// Buffer size = concurrency limit to allow provider goroutines to complete
+	// without blocking on sender
+	bufferSize := env.VirtualMachinesMaxConcurrentVsockConnections.IntegerSetting()
+	reportChan := make(chan *v1.IndexReport, bufferSize)
```

Comment:
One issue that may be a bit tricky to solve _properly_ - we must close this channel :)

Reply:
As agreed, let's leave this as TODO here (https://github.com/stackrox/stackrox/pull/18001/commits/f36d1e52ed7134e5174e92e604e56ef6bda1cb17) and implement a fix in a different PR (proposal where the producer closes the channel: https://github.com/stackrox/stackrox/pull/18017)